components/net/src/dispatcher/mod.rs

```diff
@@ -12,41 +12,63 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+pub mod prelude;
+
+use std::default::Default;
 use std::fmt;
 use std::result;
-use std::sync::{mpsc, Arc, RwLock};
+use std::sync::{Arc, RwLock};
+use std::sync::mpsc::SyncSender;
 
 use protobuf::parse_from_bytes;
 use zmq;
 
 use config::DispatcherCfg;
 use server::Envelope;
 
-pub type MessageHandler<T> = Fn(&mut Envelope) -> result::Result<(), T>;
+/// Function signature for dispatch handlers.
+pub type MessageHandler<T> = Fn(&mut Envelope) -> Result<(), T>;
+
+/// Apply to a struct containing worker state that will be passed as a mutable reference on each
+/// call of `dispatch()` to an implementer of `Dispatcher`.
+pub trait DispatcherState {
+    fn is_initialized(&self) -> bool;
+}
 
 /// Dispatchers connect to Message Queue Servers
 pub trait Dispatcher: Sized + Send {
     type Config: Send + Sync + DispatcherCfg;
     type Error: Send + From<zmq::Error> + fmt::Display;
-    type State;
+    type InitState: Clone + Send + Into<Self::State>;
+    type State: DispatcherState;
+
+    fn dispatch(message: &mut Envelope,
+                socket: &mut zmq::Socket,
+                state: &mut Self::State)
+                -> result::Result<(), Self::Error>;
 
+    /// Address to the ZeroMQ message queue that dispatcher workers will consume from.
     fn message_queue() -> &'static str;
 
-    // JW TODO: This should take something that impelements an "application config" trait
     fn new(config: Arc<RwLock<Self::Config>>) -> Self;
 
     fn context(&mut self) -> &mut zmq::Context;
 
-    fn dispatch(message: &mut Envelope,
-                socket: &mut zmq::Socket,
-                state: &mut Self::State)
-                -> result::Result<(), Self::Error>;
-
-    fn init(&mut self) -> result::Result<(), Self::Error> {
-        Ok(())
+    /// Callback to perform dispatcher initialization.
+    ///
+    /// The default implementation will take your initial state and convert it into the actual
+    /// state of the worker. Override this function if you need to perform additional steps to
+    /// initialize your worker state.
+    fn init(&mut self, init_state: Self::InitState) -> result::Result<Self::State, Self::Error> {
+        Ok(init_state.into())
     }
 
-    fn start(mut self, rz: mpsc::SyncSender<()>) -> result::Result<(), Self::Error> {
+    fn start(mut self, rz: SyncSender<()>, mut state: Self::State) -> Result<(), Self::Error> {
+        // Debug assert because it's only necessary to perform this check during development. It
+        // isn't possible for the state to not be initialized at runtime unless the developer
+        // wrongfully implements the `init()` callback or omits an override implementation where
+        // the default implementation isn't enough to initialize the dispatcher's state.
+        debug_assert!(state.is_initialized(), "Dispatcher state not initialized!");
```

Comment:
This is a developer error that should immediately be apparent and seen when running the application in a dev build. I've chosen `debug_assert` because I want people to know the source of their mistake, but I don't want this check to be done in optimized builds